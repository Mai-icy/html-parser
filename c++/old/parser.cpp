#include "parser.h"

inline Element::Element(const std::string &tag_txt) : parent_(nullptr)
{
    parse_attribute(tag_txt);
}
void Element::parse_attribute(const std::string &tag_txt)
{
    using namespace std;
    tag_ = get_tag_type(tag_txt);
    regex words_regex("(\\S*)[\\s]*=[\\s]*['\"]([^>\\s]*)['\"]");
    auto words_begin = sregex_iterator(tag_txt.cbegin(), tag_txt.cend(), words_regex);
    auto words_end = sregex_iterator();
    for (sregex_iterator k = words_begin; k != words_end; ++k)
    {
        smatch match = *k;
        string attr = match.str(1);
        string value = match.str(2);
        attribute_map[attr] = value;
        // cout << attr << ":的值:" << value << endl;
    }
}
std::string Element::get_tag_type(const std::string &tag_text)
{
    using namespace std;
    regex type_reg("([^=\\s]*)");
    smatch result;
    auto ret = regex_search(tag_text.cbegin(), tag_text.cend(), result, type_reg);
    return ret ? result.str(1) : "";
}
void Element::show()
{
    std::cout << "标签类型:" << tag_ << " 内容：" << value_ << std::endl;
}

HtmlFile::HtmlFile(const std::string &path)
{
    using namespace std;
    ofstream out_file;
    
}

void parse_ml(const std::string &ori_text)
{
    using namespace std;
    regex reg("<([^>]*)>");
    stack<pair<string, string::const_iterator>> parse_stack;

    auto words_begin = sregex_iterator(ori_text.cbegin(), ori_text.cend(), reg);
    auto words_end = sregex_iterator();
    cout << "开始" << endl;
    for (sregex_iterator k = words_begin; k != words_end; ++k)
    {
        smatch match = *k;
        string tag_text = match.str(1);

        string tag_type;
        if (tag_text[0] == '/')
        {
            string new_text(tag_text.cbegin() + 1, tag_text.cend());
            tag_type = get_tag_type(new_text);

            while (true)
            {
                auto top_pair = parse_stack.top();
                parse_stack.pop();
                string top_text = top_pair.first;
                string top_type = get_tag_type(top_text);
                if (top_type == tag_type)
                {
                    cout << "弹出" << top_type << endl;
                    string value(top_pair.second, ori_text.cbegin() + match.position());
                    Element ele(top_text);
                    break;
                }
                else
                {
                    Element ele(top_text);
                    cout << tag_text << endl;
                }
            }
        }
        else
        {
            string::const_iterator end_iter = ori_text.cbegin() + match.position() + tag_text.length() + 2;
            cout << "压入" << get_tag_type(tag_text) << endl;
            parse_stack.push(pair(tag_text, end_iter));
        }
    }
}

void parse_mml(const std::string &ori_text)
{
    using namespace std;

    stack<pair<Element, string::const_iterator>> element_stack;
    stack<Element> layer_stack;

    regex reg("<([^>]*)>");
    stack<pair<string, string::const_iterator>> parse_stack;
    auto words_begin = sregex_iterator(ori_text.cbegin(), ori_text.cend(), reg);
    auto words_end = sregex_iterator();

    bool is_last_finish = true;

    for (sregex_iterator k = words_begin; k != words_end; ++k)
    {
        smatch match = *k;
        string tag_text = match.str(1);
        string tag_type;

        if (tag_text[0] == '/')
        {
            // 标签结束标识
            string new_text(tag_text.cbegin() + 1, tag_text.cend());
            tag_type = get_tag_type(new_text);
        }
        else
        {
            // 标签添加标识
            string::const_iterator end_iter = ori_text.cbegin() + match.position() + tag_text.length() + 2;
            Element temp(tag_text);
            if (is_last_finish)
            {
            }
            else
            {
                Element last_ele = element_stack.top().first;
                layer_stack.push(last_ele);
                temp.set_parent(&last_ele);
            }
            is_last_finish = false;
        }
    }
}

std::string read_html(const std::string &html_path)
{
    using namespace std;
    ifstream html_file(html_path, ios::in);
    string res;
    string temp;
    if (!html_file.is_open())
    {
        cout << "文件打开失败" << endl;
    }
    while (getline(html_file, temp))
    {
        res += temp;
    }
    return res;
}

int main()
{
    using namespace std;

    string test{"123<xml>value</xml>456<s></s>"};

    string content = read_html("page.html");
    parse_ml(content);

    return 0;
}