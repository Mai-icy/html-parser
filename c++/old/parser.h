#include <iostream>
#include <regex>
#include <string>
#include <map>
#include <vector>
#include <stack>
#include <fstream>

class Element
{
private:
    Element *parent_;
    std::string tag_;
    std::string value_;
    std::vector<Element *> derive_ele_vector;
    std::map<std::string, std::string> attribute_map;
    std::string get_tag_type(const std::string &tag_text);
    void parse_attribute(const std::string &tag_txt);
    void set_parent(Element *parent_ele) { parent_ = parent_ele; };
    void add_derive_ele(Element *new_ele) { derive_ele_vector.push_back(new_ele); };

public:
    Element(const std::string &tag_txt);
    ~Element();
    const Element *parent() const { return parent_; };
    const std::vector<Element *> &derive_elements() const { return derive_ele_vector; };
    std::string tag() const { return tag_; };
    std::string get_attribute(std::string key) const;
    void show();

    friend HtmlFile;
};

std::string get_tag_type(const std::string &tag_text);
std::string::const_iterator search_element(const std::string &text);
std::string read_html(const std::string &html_path);
void parse_ml(const std::string &ori_text);

class HtmlFile
{
private:
    Element top_ele;
    void main_parse(const std::string &ori_text);

public:
    HtmlFile(const std::string &path);
};
