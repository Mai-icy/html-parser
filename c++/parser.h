#include <iostream>
#include <regex>
#include <string>
#include <map>
#include <vector>
#include <stack>
#include <fstream>

class Element
{
private:
    Element *parent_;
    std::string tag_;
    std::string value_;
    std::map<std::string, std::string> attribute_map;
    std::vector<Element *> derive_ele_vector;
    void parse_attribute(const std::string &tag_txt);

public:
    Element(const std::string &tag_txt, const std::string &value);
    void add_derive_ele(Element *new_ele) { derive_ele_vector.push_back(new_ele); };
    void set_parent(Element *parent_ele) { parent_ = parent_ele; };
    const Element *parent() const { return parent_; };
    const std::vector<Element *> derive_elements() const;
    std::string tag() const {return tag_;};
    std::string get_attribute(std::string key) const;
    void show();
};

std::string get_tag_type(const std::string &tag_text);
std::string::const_iterator search_element(const std::string &text);
std::string read_html(const std::string &html_path);
void parse_ml(const std::string &ori_text);
