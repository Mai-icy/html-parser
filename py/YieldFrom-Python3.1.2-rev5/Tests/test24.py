#
#   Test parser module
#

import parser

def test(src):
	st1 = parser.suite(src)
	tup1 = st1.totuple()
	print(tup1)
	st2 = parser.sequence2st(tup1)
	tup2 = st2.totuple()
	print(tup1 == tup2)

test("yield from 1")
test("f(yield from 1)")
