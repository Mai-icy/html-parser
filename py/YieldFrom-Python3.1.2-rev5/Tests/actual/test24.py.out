(257, (268, (269, (270, (275, (279, (336, (1, 'yield'), (337, (1, 'from'), (301, (305, (306, (307, (308, (310, (311, (312, (313, (314, (315, (316, (317, (318, (319, (2, '1'))))))))))))))))))))), (4, ''))), (0, ''))
True
(257, (268, (269, (270, (271, (326, (301, (305, (306, (307, (308, (310, (311, (312, (313, (314, (315, (316, (317, (318, (319, (1, 'f')), (321, (7, '('), (329, (330, (338, (1, 'yield'), (1, 'from'), (301, (305, (306, (307, (308, (310, (311, (312, (313, (314, (315, (316, (317, (318, (319, (2, '1'))))))))))))))))))), (8, ')'))))))))))))))))))), (4, ''))), (0, ''))
True
